### ISSUE TITLE
Changing how uninstalling modules works

### ISSUE THREAD

**Issue by arjunk2004** [backdrop/backdrop-issues#3939]
When you uninstall a module, it only deletes the data. I think that uninstall should remove the module entirely (from the filesystem and the ‘List Modules’ page) and there should be a ‘Clear Data’ tab or button where we clear the data for a module. I think this would be more intuitive and similar to what other software does.

**Comment by klonos** [comment]
(repeating my comments from Gitter)

@arjunk2004 contributed modules/themes have always needed to be manually removed from the filesystem in Drupal. Now that we have the Project Installer in core, perhaps it does make sense to allow people to be able to uninstall and remove the modules completely (also from the file system). Having said that, core modules should not be allowed to be removed, so these would always show up in the list of modules.

Others on Gitter have expressed concerns about security...

**Comment by klonos** [comment]
@serundeputy 03:08
>i've been trying to let the concept sink in; but my intuition is telling me this is a security concern

@klonos 03:11
>@serundeputy how so? Is it because of folder permissions?

@serundeputy 03:12
>yes; you'd be alllowing the web user to delete stuff from disk; but we allow them to write to disk with project browser :shrug:
@serundeputy i don't turn on project installer on any of my sites; and opt for the stricter file permissions

@herbdool 03:15
>Automatically removing code when a module is uninstalled would be problematic. Some people intend to reinstall it. It could be custom code and don't have a copy anywhere else. Some people use version control and/or lock the files so can't just delete files

@klonos 03:17
>Yeah, I was about to say that we already allow writing to /modules, /themes, /layouts ...and with automatic security updates in core, also /core (actually the last is replacing the entire /core dir IIGTR).
>...with #3146 we could have it so that core and custom modules do not have the option to be removed.

**Comment by findlabnet** [comment]
Web-server user with filesystem management rights? Sorry, but worst thing cannot be found.

**Comment by klonos** [comment]
@findlabnet I know, but we already allow some filesystem permissions in order for the Project Browser to work OOTB.

**Comment by findlabnet** [comment]
Not on my servers. 
@klonos - errors can be fixed, you know ;-)

**Comment by arjunk2004** [comment]
Perhaps then we just have a reminder when it’s uninstalled telling them that data has been removed but not the module and how to remove the module?

**Comment by findlabnet** [comment]
@arjunk2004 any module can have place in the filesystem, but never installed, even core modules.

**Comment by arjunk2004** [comment]
@findlabnet I'm not quite sure what you mean. What do you mean by a module is never installed.

**Comment by findlabnet** [comment]
@arjunk2004 - core module "Book" for example.

**Comment by stpaultim** [comment]
@arjunk2004 - I believe that what @findlabnet means is that many of us don't really think about a module as being `installed` just because it is listed here: admin/modules/list

I often don't think of a module as being installed until I've checked the box and clicked on the configure button at the bottom of the page. 

The `Book` and `Contact` module are available to me with Drupal out of the box, but I don't think of them as installed. 

![Modules___Dashboard](https://user-images.githubusercontent.com/3144571/61611801-61221580-ac22-11e9-8cf5-dd001a4e27a7.png)

I understand, that this conflicts with the terminology we use in the UI. 

Based upon the `install` page - admin/modules/install it would appear that once a module has been loaded into the file system that it is installed, whether or not the box is checked and it has been enabled. 

![Install_modules___Dashboard](https://user-images.githubusercontent.com/3144571/61611966-dbeb3080-ac22-11e9-987d-1833b32d5578.png)

I think that the problem comes from the fact that there are really only two steps to add a module and 3 steps to remove it. 

To add:

1) Copy to file system
2) Check box and configure (this enables it AND adds configures the database/config)

To remove

1) Disable (just disables the functionality without removing entirely from database)
2) Uninstall (Removes all traces from database/config)
3) Delete from file system

**Comment by arjunk2004** [comment]
I created a PR so it tells you to delete the module from the file system when you "uninstall" the module. Please tell me what y'all think of this...

**Comment by olafgrabienski** [comment]
@arjunk2004 Thanks for the PR which adds the sentence `Note: The module data has been reset but to delete it, you must do so through the filesystem.` to the success message on the *Uninstall* page.

In my opinion the confirmation could be improved using two message types:

- Success message: `The selected modules have been uninstalled.`
- Info message: `If you want to delete an uninstalled module, remove it from your filesystem.`

**Comment by findlabnet** [comment]
One more time: what about uninstalling unused core module - message should propose remove it from filesystem too?

**Comment by olafgrabienski** [comment]
You're right, we should show the info message only if a non core module has been disabled. Don't know if such a distinction is possible.

**Comment by klonos** [comment]
> In my opinion the confirmation could be improved using two message types:

@olafgrabienski you beat me to it, I had written up a responce, but then my browser crashed on me. As part of that responce, I created this mockup:

<img width="1048" alt="Screen Shot 2019-08-13 at 10 02 07 pm" src="https://user-images.githubusercontent.com/2423362/63106390-66595280-bf8b-11e9-93b4-d4fcd3365464.png">

My info message is rather wordy, so perhaps we can consider shortening it, or have a link to online documentation on b.org. But yeah, mixing a success message with a notice/warning/info does not seem right to me either.

**Comment by klonos** [comment]
...the text I came up with was generic/info, and my suggestion was to have it shown all the time on the "Uninstall modules" page; not only upon an uninstallation.

As for whether we can distinguish between core vs contrib/custom, this should be doable, since core modules are within `/core`, so we could be checking the path module.

**Comment by stpaultim** [comment]
Just sharing this note from a user in the forum:

> For me it seems logical to remove the contributed module files when uninstalling it but I understand all care should be taken regarding security issues.

https://forum.backdropcms.org/forum/uninstalled-modules-are-not-removed

**Comment by stpaultim** [comment]
Here is another post in the forum:

> Oh okay, it is just weird that I can install and system can download but can not delete.

https://forum.backdropcms.org/comment/2836#comment-2836

**Comment by philsward** [comment]
Agree we need this in a bad way.

I don't like the idea of deleting and uninstalling at the same time which means we need a new tab to deal with uninstalled modules that are still registered to the system. I too am not a fan of the increased workflow as @stpaultim pointed out, but I have a feeling it can't be avoided, but can be lessened.

I would propose having two buttons at the bottom of the Uninstall page

[Uninstall] [Uninstall and Delete]

Add another tab for Delete Modules and list any registered modules that are uninstalled but haven't been deleted and allow someone to check off the ones they want deleted. Maybe add a locking mechanism so they can't be deleted without being unlocked in the case of custom contrib modules?

@findlabnet 
>Not on my servers.

So... what about the temp files created when a file is uploaded? What about cron purging cache files? Drupal / BD already allow the deletion / movement of files within the file system, so deleting modules isn't as much as a security issue as what is already happening behind the scenes.

Maybe it would be more secure to set a flag in the DB to delete the module and let cron handle the deletion instead of executing it directly? I don't know enough about this side of it to know what's best but it might be an option.

**Comment by domaingood** [comment]
I dot agree with this for security issue. no need to delete bottom. If need delete they can via from file manager via FTP or other mathod via server.
I would propose having two buttons at the bottom of the Uninstall page

[Uninstall] [Uninstall and Delete]

**Comment by findlabnet** [comment]
@philsward 
> So... what about the temp files

There are only two such places: the system directory "tmp" and "files" (in Drupal / BD under the control of a web server).
But open the entire file system to the web server?
And no, I don't use the Project Installer at all.
This is just my opinion / decision.
Although I could be wrong.

**Comment by philsward** [comment]
@findlabnet no I think your concerns are warranted and if you want to lock your server down, that is your call.

I think this problem can be approached from a security and a good UX standpoint though. We just need to figure out what that will look like.

**Comment by philsward** [comment]
I don't know if there is an issue out to have the project installer automatically download the files into a ./modules/contrib folder, but if not I think we should get that going first, then we can lock down the ./modules/contrib folder to only allow the modules that are downloaded by the project installer to be queued for deletion. This will help mitigate (most?) security risks and concerns.

I get why there isn't a "contrib" folder in ./modules already due to simplicity, but I really think we shouldn't pander just to "new people" and go ahead and make it a practice that contributed modules go into the ./modules/contrib folder and anything else either goes in ./modules or ./custom. Maybe the modules folder is blank on fresh install but the project installer puts them into the ./contrib folder?

Drush for Drupal (assuming Brush for Backdrop does too) automatically downloads into a ./contrib folder if it exists. Having a single folder that is only accessible by the project installer would be the best way to lock things down and keep custom code completely separate from accidental deletion.

There would need to be some sort of indicator that lets the site builder know something like: "This module is not within the contrib folder and can't be queued for deletion." This would also allow site admins to "lock" certain contrib projects from being deleted accidentally by a lesser admin who has permission to add and remove modules, but really shouldn't be able to remove "that one particular module for their own safety".

Just some thoughts.
